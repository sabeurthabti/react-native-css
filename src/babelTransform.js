import RNC from './index';
import path from 'path';
import {spawn} from 'child_process';
const css = new RNC();

function runWatcher(input, output) {
  spawn('node', ['./watcher.js', input, output], {detached: true, cwd: __dirname});
}

function startTransform(input, output) {

  css.parse({
    input,
    output,
    useInheritance: true
  });

  if (process.env.BABEL_ENV === 'development') {
    runWatcher(input, output);
  }
  return true;
}

//

export default function ({ types: t }) {
  return {
    visitor: {
      ImportDeclaration (transformPath, {file}) {
        let resolvePath = transformPath.node.source.value;
        if (resolvePath.startsWith('css!')) {
          resolvePath = resolvePath.substr(4);
          let name = resolvePath.replace(/\.\.\/|\.\//g, '').replace(/\//g, '_').split('.')[0];
          let absolutePath = path.resolve(path.dirname(file.opts.filename), resolvePath),
              relativePath = `${path.dirname(absolutePath)}/_transformed/${name}.js`;
          startTransform(absolutePath, path.resolve(relativePath));

          let expression = (t.callExpression(t.memberExpression(t.callExpression(t.identifier('require'), [t.stringLiteral('react-native-css')]), t.identifier('register')), [t.callExpression(t.identifier('require'), [path.resolve(relativePath)])]));
          transformPath.replaceWith(expression);
        }
      }
    }
  };
}
