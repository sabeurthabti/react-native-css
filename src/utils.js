import fs from "fs-extra";
export default class Utils {

  static arrayContains(value, arr) {
    for (var i = 0; i < arr.length; i++) {
      if (value === arr[i]) {
        return true;
      }
    }
    return false;
  }

  static filterArray(targetArray, excludeArray) {
    return targetArray.filter(value=>excludeArray.indexOf(value) === -1);
  }

  static clean(string) {
    return string.replace(/\r?\n|\r/g, "");
  }

  static readFile(file) {
    return fs.readFileSync(file, "utf8");
  }

  static outputReactFriendlyStyle(style, outputFile, prettyPrint, literalObject) {
    var indentation = prettyPrint ? 4 : 0;
    var jsonOutput = JSON.stringify(style, null, indentation);
    var output = "module.exports = ";
    output += (literalObject) ? `${jsonOutput}` : `require('react-native').StyleSheet.create(${jsonOutput});`;
    // Write to file
    if (outputFile) {
      fs.ensureFileSync(outputFile);
      fs.writeFileSync(outputFile, output);
    }
    return output;
  }

  static contains(string, needle) {
    var search = string.match(needle);
    return search && search.length > 0;
  }
}
