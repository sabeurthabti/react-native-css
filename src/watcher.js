import 'babel-polyfill';
import RNC from './index';
import fs from 'fs';
import request from 'request';
import chokidar from 'chokidar';

const css = new RNC();

/**
 * Checks if the packager is running, if not there really isn't point of keeping this thing alive.
 * @return {Promise}
 */
function isServerRunning() {
  return new Promise(resolve=>request('http://localhost:8081/status', (err, response, body)=> {
    resolve(body === "packager-status:running");
  }));
}

async function run() {
  let input  = process.argv[2],
    output = process.argv[3];

  if (!input || !output) {
    return;
  }

  //check pid
  let name    = input.replace(/[^0-9a-z]/g, ''),
    pidFile = `/tmp/rnc_watch_${name}.pid`;

  function check() {
    try {
      return process.kill(~~fs.readFileSync(pidFile, 'utf8'), 0);
    }
    catch (e) {
      return e.code === 'EPERM';
    }
  }

  if (check()) {
    return process.exit(0);
  }

  fs.writeFileSync(pidFile, process.pid, 'utf8');

  //check if server is up
  if (!(await isServerRunning())) {
    return process.exit(0);
  }

  chokidar.watch(input, {
    ignored: /[\/\\]\./, persistent: true
  }).on('change', function () {
    css.parse({input, output, useInheritance: true});
  });

  setInterval(async ()=> {
    if (!(await isServerRunning())) {
      process.exit(0);
    }
  }, 10000);

}


run();
