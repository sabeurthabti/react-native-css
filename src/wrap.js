import React,{Component} from "react";
import css,{matchingRules} from "./css";
import shallowCompare from "react/lib/shallowCompare";
import hoistNonReactStatics from "hoist-non-react-statics";


let isWebApp = false;
try {
  //Check if this is actually in the browser, this is probably leveraging react-native-web and allows for sharing
  // styled components between web and native apps.
  /* eslint no-undef: "off" */
  if (typeof window.document.getElementById === "function") {
    isWebApp = true;
  }
} catch (e) {
  //looks like pure React Native
}

function normalizeClassNames(classNames) {
  let classes = [];
  if (typeof classNames === "string") {
    classes = classNames.split(" ").filter(a=>a);
  } else if (classNames instanceof Array) {
    classNames.forEach(name=> {
      classes.push(...normalizeClassNames(name));
    });
  }
  return classes;
}

function indexOf(instance, children) {
  let index = 0;
  for (let key in children) {
    if (children[key]._mountOrder === instance._reactInternalInstance._mountOrder) {
      return index;
    }
    index++;
  }
  return -1;
}

const pathCache = {};

/**
 * Takes the passed component and wraps it in a styled proxy.  All props are passed through.
 * Path information is handled via Context, so be careful not to override the 'path' context value.
 * @param {string} name Name that will be used in CSS selectors, case insensitive.
 * @param {Component} WrappedComponent
 * @return {Component}
 */
export default function wrap(name, WrappedComponent) {
  if (isWebApp) {
    //Since the whole point of this is to mimic CSS rules from the browser, just simply return the component in a web
    // environment, no wrapping necessary.
    return WrappedComponent;
  }
  let StyledComponent = class extends Component {

    componentWillReceiveProps(nextProps) {
      if (shallowCompare(this, nextProps)) {
        this.cssPathKey = null;
      }
    }

    //This is the magic right here.
    getChildContext() {
      let self = this;
      return {
        get cssPath() {
          if (!self.cssPath) {
            self.createPath(self.props);
          }
          return self.cssPath;
        },
        get cssPathKey() {
          if (!self.cssPath) {
            self.createPath(self.props);
          }
          return self.cssPathKey;
        }
      };
    }

    get pathKey() {
      if (!this.cssPathKey || this.context.cssPathKey !== this._lastPathKey) {
        this.createPath(this.props);
        this._lastPathKey = this.context.cssPathKey;
      }
      return this.cssPathKey;
    }

    get styles() {
      if (!this.style || this._lastKey !== this.pathKey) {
        if (!this.cssPath) {
          this.createPath(this.props);
        }
        let style = matchingRules(this.cssPath, this.cssPathKey);
        this.style = this.props.style ? css(style, this.props.style) : style;
        this._lastKey = this.pathKey;
      }
      return this.style;
    }

    setNativeProps(nativeProps) {
      this._root && this._root.setNativeProps(nativeProps);
    }

    shouldComponentUpdate(nextProps) {
      return shallowCompare(this, nextProps);
    }

    /**
     * This updates the path and style with the latest props
     * @param props
     */
    createPath(props) {
      let index = -1,
        count = 1;
      if (this._reactInternalInstance && this._reactInternalInstance._hostParent) {
        let children = this._reactInternalInstance._hostParent._renderedChildren;
        count = Object.keys(children).length;
        index = indexOf(this, children);
      }
      let element = {
        e: name.toLowerCase(),
        c: normalizeClassNames(props.className),
        //Maybe in the future, but is it worth the performance hit?
//        p: props,
        i: index,
        f: index === 0,
        l: count === index + 1
      };
      let key = `${this.context.cssPathKey || ""}>${element.e}.${element.c.join(".")}:${element.i || ""}:${element.f || ""}:${element.l || ""}`;
      if (this.cssPathKey !== key) {
        //Create the path or use the cache.  Since this can be called thousands of times, the cache reduces the more
        // expensive array generations and reduces memory consumption by resusing arrays. Because of this it is
        // important that the arrays are treated as immutable.
        this.cssPath = pathCache[key] || (pathCache[key] = (this.context.cssPath || [{e: "root"}]).concat([element]));
        this.cssPathKey = key;
      }
    }

    render() {
      return <WrappedComponent {...this.props} ref={ref=>this._root = ref}
                                               style={this.props.style?[this.styles,this.props.style]:this.styles}/>;
    }
  };

  hoistNonReactStatics(StyledComponent, WrappedComponent);
  StyledComponent.WrappedComponent = WrappedComponent;
  StyledComponent.displayName = name;
  StyledComponent.childContextTypes = {
    cssPath: React.PropTypes.array,
    cssPathKey: React.PropTypes.string
  };
  //Make sure we are getting our parent's path
  StyledComponent.contextTypes = {cssPath: React.PropTypes.array, cssPathKey: React.PropTypes.string};
  return StyledComponent;
}
