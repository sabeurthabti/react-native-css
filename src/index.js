import ParseCSS from 'css-parse';
import toCamelCase from 'to-camel-case';
import utils from './utils';
import inheritance,{numeric} from './inheritance';

export default class ReactNativeCss {

  parse({input, output, prettyPrint = false, literalObject = false, useInheritance = false}) {
    if (!input) {
      throw new Error('An input file is required.');
    }
    let data;
    if (utils.contains(input, /scss/)) {
      let {css} = require('node-sass').renderSync({
        file: input,
        outputStyle: 'compressed'
      });
      data = css.toString();
    } else {
      data = utils.readFile(input);
    }

    let styleSheet = this.toJSS(data, useInheritance);
    if (output) {
      utils.outputReactFriendlyStyle(styleSheet, output, prettyPrint, literalObject);
    }
    return styleSheet;
  }

  toJSS(stylesheetString, useInheritance = false) {
    const directions = ['top', 'right', 'bottom', 'left'];
    const changeArr = ['margin', 'padding', 'border-width', 'border-radius'];
    const numberize = utils.filterArray(['width', 'height', 'font-size', 'line-height'].concat(directions), useInheritance ? numeric : []);
    //special properties and shorthands that need to be broken down separately
    const specialProperties = {};
    ['border', 'border-top', 'border-right', 'border-bottom', 'border-left'].forEach(name=> {
      specialProperties[name] = {
        regex: /^\s*([0-9]+)(px)?\s+(solid|dotted|dashed)?\s*([a-z0-9#,\(\)\.\s]+)\s*$/i,
        map: {
          1: `${name}-width`,
          3: name == 'border' ? `${name}-style` : null,
          4: `${name}-color`
        }
      };
    });

    directions.forEach((dir) => {
      numberize.push(`border-${dir}-width`);
      changeArr.forEach((prop) => {
        numberize.push(`${prop}-${dir}`);
      });
    });

    //map of properties that when expanded use different directions than the default Top,Right,Bottom,Left.
    const directionMaps = {
      'border-radius': {
        'Top': 'top-left',
        'Right': 'top-right',
        'Bottom': 'bottom-right',
        'Left': 'bottom-left'
      }
    };

    //Convert the shorthand property to the individual directions, handles edge cases, i.e. border-width and
    // border-radius
    function directionToPropertyName(property, direction) {
      let names = property.split('-');
      names.splice(1, 0, directionMaps[property] ? directionMaps[property][direction] : direction);
      return toCamelCase(names.join('-'));
    }

    // CSS properties that are not supported by React Native
    // The list of supported properties is at
    // https://facebook.github.io/react-native/docs/style.html#supported-properties
    const unsupported = ['display'];

    let {stylesheet} = ParseCSS(utils.clean(stylesheetString));

    let JSONResult = {};

    for (let rule of stylesheet.rules) {
      if (rule.type !== 'rule') {
        continue;
      }

      for (let selector of rule.selectors) {
        if (!useInheritance) {
          selector = selector.replace(/\.|#/g, '').trim();
        }

        let styles = (JSONResult[selector] = JSONResult[selector] || {});

        for (let declaration of rule.declarations) {

          if (declaration.type !== 'declaration') {
            continue;
          }

          let value = declaration.value;
          let property = declaration.property;

          if (specialProperties[property]) {
            let special = specialProperties[property],
              matches = special.regex.exec(value);
            if (matches) {
              if (typeof special.map === 'function') {
                special.map(matches, styles, rule.declarations);
              } else {
                for (let key in special.map) {
                  if (matches[key] && special.map[key]) {
                    rule.declarations.push({
                      property: special.map[key],
                      value: matches[key],
                      type: 'declaration'
                    });
                  }
                }
              }
              continue;
            }
          }

          if (utils.arrayContains(property, unsupported)) {
            continue;
          }

          if (utils.arrayContains(property, numberize)) {
            styles[toCamelCase(property)] = parseFloat(value.replace(/px|\s*/g, ''));
          }

          else if (utils.arrayContains(property, changeArr)) {

            var values = value.replace(/px/g, '').split(/[\s,]+/);

            values.forEach(function (value, index, arr) {
              arr[index] = parseInt(value);
            });

            var length = values.length;

            if (length === 1) {

              styles[toCamelCase(property)] = values[0];

            }

            if (length === 2) {

              for (let prop of ['Top', 'Bottom']) {
                styles[directionToPropertyName(property, prop)] = values[0];
              }

              for (let prop of ['Left', 'Right']) {
                styles[directionToPropertyName(property, prop)] = values[1];
              }
            }

            if (length === 3) {

              for (let prop of ['Left', 'Right']) {
                styles[directionToPropertyName(property, prop)] = values[1];
              }

              styles[directionToPropertyName(property, 'Top')] = values[0];
              styles[directionToPropertyName(property, 'Bottom')] = values[2];
            }

            if (length === 4) {
              ['Top', 'Right', 'Bottom', 'Left'].forEach(function (prop, index) {
                styles[directionToPropertyName(property, prop)] = values[index];
              });
            }
          }
          else {
            if (!isNaN(declaration.value) && property !== 'font-weight') {
              declaration.value = parseFloat(declaration.value);
            }

            styles[toCamelCase(property)] = declaration.value;
          }
        }
      }
    }
    return useInheritance ? inheritance(JSONResult) : JSONResult;
  }
}
