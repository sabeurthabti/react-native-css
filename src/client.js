import ReactNative from "react-native";
import css,{register,unregister} from "./css";
import wrap from "./wrap";

//These are all the components in react native to wrap with the styled proxy component
const componentsToWrap = ["ActivityIndicator",
  "ActivityIndicatorIOS",
  "ART",
  "DatePickerIOS",
  "DrawerLayoutAndroid",
  "Image",
  "ImageEditor",
  "ImageStore",
  "KeyboardAvoidingView",
  "ListView",
  "MapView",
  "Modal",
  "Navigator",
  "NavigatorIOS",
  "Picker",
  "PickerIOS",
  "ProgressBarAndroid",
  "ProgressViewIOS",
  "ScrollView",
  "SegmentedControlIOS",
  "Slider",
  "SliderIOS",
  "SnapshotViewIOS",
  "Switch",
  "RecyclerViewBackedScrollView",
  "RefreshControl",
  "StatusBar",
  "SwipeableListView",
  "SwitchAndroid",
  "SwitchIOS",
  "TabBarIOS",
  "Text",
  "TextInput",
  "ToastAndroid",
  "ToolbarAndroid",
  "Touchable",
  "TouchableHighlight",
  "TouchableNativeFeedback",
  "TouchableOpacity",
  "TouchableWithoutFeedback",
  "View",
  "ViewPagerAndroid",
  "WebView"];


const StyledComponents = {};
let wrappedComponents = {};
//wrap all the UI components in React Native
componentsToWrap.forEach(name=> {
  Object.defineProperty(StyledComponents, name, {
    get(){
      return wrappedComponents[name] || (wrappedComponents[name] = wrap(name, ReactNative[name]));
    }
  });
});

//Proxy all other APIs, etc. so it can be a drop-in replacement for import {...} from 'react-native.  All properties
// use a getter so APIs are only initialized when explicitly imported
Object.keys(ReactNative).forEach(name=> {
  if (componentsToWrap.indexOf(name) === -1) {
    Object.defineProperty(StyledComponents, name, {
      get(){
        return ReactNative[name];
      }
    });
  }
});

//Add in all the utility functions
StyledComponents.unregister = unregister;
StyledComponents.register = register;
StyledComponents.css = css;
StyledComponents.wrap = wrap;

module.exports = StyledComponents;
