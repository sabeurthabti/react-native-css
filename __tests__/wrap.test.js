import React, {Component} from 'react';
import {View,Text} from 'react-native';

import renderer from 'react-test-renderer';

import RNC from '../src/index';
import {register} from '../src/css';
import wrap from '../src/wrap';
const css = new RNC();

global.Promise = require.requireActual('promise');


describe('wrap', ()=> {
  it('should pass all properties to wrapped component and add styles to the view', async ()=> {
    //create the style object
    let styles = css.parse({
      input: './__tests__/fixtures/style-inherit.css',
      useInheritance: true
    });

    //register the global sheet
    register(styles);

    //define the wrapped component
    const TestComponent = wrap('TestComponent', class extends Component {
      render() {
        return <View {...this.props}/>;
      }
    });

    //render it
    const tree = renderer.create(
      <TestComponent className="direct" prop1="1" prop2="3" prop4="4"/>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

});




