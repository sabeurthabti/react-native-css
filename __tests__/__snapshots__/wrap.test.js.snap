// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrap should pass all properties to wrapped component and add styles to the view 1`] = `
<View
  className="direct"
  prop1="1"
  prop2="3"
  prop4="4"
  style={
    Object {
      "margin": 5,
    }
  }
/>
`;
